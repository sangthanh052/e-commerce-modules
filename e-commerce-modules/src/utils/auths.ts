import type { User } from '@/types/user.type'

export const LocalStorageEventTarget = new EventTarget() //Api có sẵn trình duyệt

export const setAccessTokenToLS = (access_token: string) => {
  localStorage.setItem('access_token', access_token)
}

export const setRefreshTokentoLS = (refresh_token: string) => {
  localStorage.setItem('refresh_token', refresh_token)
}

export const clearLS = () => {
  localStorage.removeItem('access_token')
  localStorage.removeItem('refresh_token')
  localStorage.removeItem('profile')
  const clearLSEvent = new Event('clearLs') //Api có sẵn trình duyệt
  LocalStorageEventTarget.dispatchEvent(clearLSEvent)
}

export const getAccessTokenFromLS = () => localStorage.getItem('access_token') || ''
export const getRefreshTokenFromLS = () => localStorage.getItem('refresh_token') || ''

export const getprofileFromLS = () => {
  const result = localStorage.getItem('profile')
  return result ? JSON.parse(result) : null
}

export const setProfileToLS = (profile: User) => {
  const result = JSON.stringify(profile)
  localStorage.setItem('profile', result)
}
