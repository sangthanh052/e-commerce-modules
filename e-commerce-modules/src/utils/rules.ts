import * as yup from 'yup'
import * as z from 'zod'

const handleConfirmPasswordYup = (refString: string) => {
  return yup
    .string()
    .required('Nhập lại password là bắt buộc')
    .min(6, 'Độ dài từ 6 - 160 ký tự')
    .max(160, 'Độ dài từ 6 - 160 ký tự')
    .oneOf([yup.ref(refString)], 'Nhập lại password không khớp')
}

export const RegisterYupSchema = yup
  .object({
    email: yup
      .string()
      .required('Email Address is required')
      .email('Email Address is Invalid')
      .min(5, 'Length from 5-160 characters')
      .max(160, 'Length from 5-160 characters')
      .matches(/^[^\s@]+@[^\s@]+\.[^\s@]+$/, 'Email must contain a dot (.) after the "@" symbol.'),

    password: yup
      .string()
      .required('Password is required')
      .min(5, 'Length from 5-160 characters')
      .max(160, 'Length from 5-160 characters'),

    confirm_password: handleConfirmPasswordYup('password')
  })
  .required()

export const RegisterZodSchema = z
  .object({
    email: z
      // .string() version moi khong can
      .email('Email Address is Invalid')
      .min(1, 'Email Address is required')
      .min(5, 'Length from 5-160 characters')
      .max(160, 'Length from 5-160 characters'),

    password: z
      .string()
      .min(1, 'Password is required')
      .min(5, 'Length from 5-160 characters')
      .max(160, 'Length from 5-160 characters'),

    confirm_password: z
      .string()
      .min(1, 'Nhập lại password là bắt buộc')
      .min(6, 'Độ dài từ 6 - 160 ký tự')
      .max(160, 'Độ dài từ 6 - 160 ký tự')
  })
  .refine((data) => data.password === data.confirm_password, {
    path: ['confirm_password'],
    message: 'Mật khẩu không khớp'
  })

export const LoginYupSchema = RegisterYupSchema.omit(['confirm_password'])
export const LoginZodSchema = RegisterZodSchema.omit({ confirm_password: true })

export type LoginYupSchemaType = yup.InferType<typeof LoginYupSchema>
export type LoginZodSchemaType = z.infer<typeof LoginZodSchema>
