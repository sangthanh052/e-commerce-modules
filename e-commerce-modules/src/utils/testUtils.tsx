import App from '@/App'
import { AppProvider, getInitialAppContext } from '@/contexts/app.context'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { screen, waitFor, type waitForOptions } from '@testing-library/dom'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { BrowserRouter } from 'react-router-dom'

export const delay = (time: number) =>
  new Promise((resolve) => {
    setTimeout(() => {
      resolve(true)
    }, time)
  })

export const logScreenImmediate = (body: HTMLElement = document.body.parentElement as HTMLElement) => {
  screen.debug(body, 99999999)
}

export const logScreenAfter = async (body: HTMLElement = document.body, options?: waitForOptions) => {
  const { timeout = 1000 } = options || {}
  await waitFor(
    () => {
      screen.debug(body, 99999999)
    },
    { timeout }
  )
}

const createWrapper = () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false
      },
      mutations: {
        retry: false
      }
    }
  })

  const Provider = ({ children }: { children: React.ReactNode }) => (
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  )
  return Provider
}

const Provider = createWrapper()

export const renderWithRouter = ({ route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route)
  const defaultValueAppContext = getInitialAppContext()

  return {
    user: userEvent.setup(),
    ...render(
      <Provider>
        <AppProvider defaultValue={defaultValueAppContext}>
          <App />
        </AppProvider>
      </Provider>,
      { wrapper: BrowserRouter }
    )
  }
}
