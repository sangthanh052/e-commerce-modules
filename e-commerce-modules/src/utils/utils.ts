import type { ErrorResponse } from '@/types/utils.type'
import axios, { AxiosError, HttpStatusCode } from 'axios'

export function isAxiosError<T>(error: unknown): error is AxiosError<T> {
  return axios.isAxiosError(error)
}

//Lỗi 422
export function isAxiosUnprocessableEntityError<FormError>(error: unknown): error is AxiosError<FormError> {
  return isAxiosError(error) && error.response?.status === HttpStatusCode.UnprocessableEntity
}

//Lỗi 401 chung
export function isAxiosUnauthorizedError<UnauthorizedError>(error: unknown): error is AxiosError<UnauthorizedError> {
  return isAxiosError(error) && error.response?.status === HttpStatusCode.Unauthorized
}

export function isAxiosExpiredTokenError<UnauthorizedError>(error: unknown): error is AxiosError<UnauthorizedError> {
  return (
    isAxiosUnauthorizedError<ErrorResponse<{ name: string; message: string }>>(error) &&
    error.response?.data?.data?.name === 'EXPIRED_TOKEN'
  )
}

export function formatCurrency(currency: number) {
  return new Intl.NumberFormat('en', { currency: 'VND', currencyDisplay: 'narrowSymbol', style: 'currency' })
    .format(currency)
    .replaceAll(',', '.')
}

export function rateSale(original: number, sale: number) {
  return Math.round(((original - sale) / original) * 100) + '%'
}

const removeSpecialCharacter = (str: string) =>
  // eslint-disable-next-line no-useless-escape
  str.replace(/!|@|%|\^|\*|\(|\)|\+|\=|\<|\>|\?|\/|,|\.|\:|\;|\'|\"|\&|\#|\[|\]|~|\$|_|`|-|{|}|\||\\/g, '')

export function generateNameId({ name, id }: { name: string; id: string }) {
  return removeSpecialCharacter(name).replace('/\s/g', '-') + `-i.${id}`
}

export const getIdFromNameId = (nameId: string) => {
  const arr = nameId.split('-i.')
  return arr[arr.length - 1] // lấy item cuối mảng
}
