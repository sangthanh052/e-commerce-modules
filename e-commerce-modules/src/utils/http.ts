/* eslint-disable @typescript-eslint/no-explicit-any */
import axios, { AxiosError, HttpStatusCode, type AxiosInstance, type InternalAxiosRequestConfig } from 'axios'
import { toast } from 'react-toastify'

import { URL_LOGIN, URL_LOGOUT, URL_REFRESH_TOKEN, URL_REGISTER } from '@/apis/auth.api'
import type { AuthResponse, RefreshTokenReponse } from '@/types/auth.type'
import type { ErrorResponse } from '@/types/utils.type'
import {
  clearLS,
  getAccessTokenFromLS,
  getRefreshTokenFromLS,
  setAccessTokenToLS,
  setProfileToLS,
  setRefreshTokentoLS
} from './auths'
import { isAxiosExpiredTokenError, isAxiosUnauthorizedError } from './utils'

export class Http {
  instance: AxiosInstance
  private accessToken: string
  private refreshToken: string
  private refreshTokenRequest: Promise<string> | null

  constructor() {
    this.accessToken = getAccessTokenFromLS()
    this.refreshToken = getRefreshTokenFromLS()
    this.refreshTokenRequest = null

    this.instance = axios.create({
      baseURL: 'https://api-ecom.duthanhduoc.com',
      timeout: 10000,
      headers: {
        'Content-Type': 'application/json',
        'expire-access-token': 10 // 10 giây
        // 'expire-refresh-token': 60 * 60 // 1 giờ
      }
    })

    this.setupInterceptors()
  }

  private setupInterceptors() {
    this.instance.interceptors.request.use(
      (config) => {
        if (this.accessToken) {
          config.headers.authorization = this.accessToken
        }
        return config
      },
      (error) => Promise.reject(error)
    )

    this.instance.interceptors.response.use(
      (response) => {
        // xử lý login, register, logout
        const { url } = response.config
        if (url === URL_LOGIN || url === URL_REGISTER) {
          const data = response.data as AuthResponse
          this.accessToken = data.data.access_token
          this.refreshToken = data.data.refresh_token
          setAccessTokenToLS(this.accessToken)
          setRefreshTokentoLS(this.refreshToken)
          setProfileToLS(data.data.user)
        } else if (url === URL_LOGOUT) {
          this.clearAuth()
        }

        return response
      },
      (error: AxiosError) => this.handleErrorResponse(error)
    )
  }

  private async handleErrorResponse(error: AxiosError) {
    const status = error.response?.status

    // chỉ toast khi không phải 422 hoặc 401
    if (![HttpStatusCode.UnprocessableEntity, HttpStatusCode.Unauthorized].includes(status as number)) {
      const data = error.response?.data as any
      const message = data?.message || error.message
      toast.error(message)
    }

    if (isAxiosUnauthorizedError<ErrorResponse<{ name: string; message: string }>>(error)) {
      const config = (error.response?.config || { headers: {} }) as InternalAxiosRequestConfig
      const { url } = config

      if (isAxiosExpiredTokenError(error) && url !== URL_REFRESH_TOKEN) {
        // Nếu this.refreshTokenRequest đã có → dùng lại Promise đang chạy.
        // Nếu chưa có → gọi handleRefreshToken() để tạo request refresh.
        // ví dụ có 5 rq thì chạy 5 lần nhưng chỉ dùng handleRefreshToken() 1 lần trong khoảng 10s.
        // Kỹ thuật này gọi là deduplication hoặc request coalescing.
        this.refreshTokenRequest =
          this.refreshTokenRequest ??
          this.handleRefreshToken().finally(() => {
            setTimeout(() => (this.refreshTokenRequest = null), 10000)
          })

        return this.refreshTokenRequest.then((access_token) =>
          this.instance({
            ...config,
            headers: { ...config.headers, authorization: access_token }
          })
        )
      }

      this.clearAuth()
      toast.error((error.response?.data as any)?.data?.message || (error.response?.data as any)?.message)
    }

    return Promise.reject(error)
  }

  private handleRefreshToken() {
    return this.instance
      .post<RefreshTokenReponse>(URL_REFRESH_TOKEN, {
        refresh_token: this.refreshToken
      })
      .then((res) => {
        const { access_token } = res.data.data
        this.accessToken = access_token
        setAccessTokenToLS(access_token)
        return access_token
      })
      .catch((error) => {
        this.clearAuth()
        throw error
      })
  }

  private clearAuth() {
    clearLS()
    this.accessToken = ''
    this.refreshToken = ''
  }
}

const http = new Http().instance
export default http
