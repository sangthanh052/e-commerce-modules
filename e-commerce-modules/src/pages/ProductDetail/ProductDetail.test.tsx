import { describe, expect, test } from 'vitest'

import { delay, renderWithRouter } from '@/utils/testUtils'

beforeAll(() => {
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: vi.fn().mockImplementation((query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: vi.fn(),
      removeListener: vi.fn(),
      addEventListener: vi.fn(),
      removeEventListener: vi.fn(),
      dispatchEvent: vi.fn()
    }))
  })
})


describe('ProductDetail', () => {
  test('Render UI ProductDetail', async () => {
    renderWithRouter({
      route: '/Điện%20Thoại%20Vsmart%20Active%203%206GB64GB%20%20Hàng%20Chính%20Hãng-i.60afb2c76ef5b902180aacba'
    })

    await delay(1000)
    expect(document.body).toMatchSnapshot()
  })
})
