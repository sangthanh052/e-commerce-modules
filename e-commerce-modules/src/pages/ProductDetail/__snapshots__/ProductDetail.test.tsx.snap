// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductDetail > Render UI ProductDetail 1`] = `
<body>
  <div>
    <script
      nonce=""
    >
      ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","light",null,["light","dark"],null,true,true)
    </script>
    <div>
      <main>
        <div />
        <div
          class="pb-5 text-sm"
        >
          <div
            class="container mx-auto"
          >
            <section
              class="grid lg:grid-cols-12"
            >
              <section
                class="lg:col-span-5 lg:p-3.5"
              >
                <div
                  class="product-image relative w-full pt-[100%] shadow"
                >
                  <img
                    alt="Điện Thoại Vsmart Active 3 6GB/64GB - Hàng Chính Hãng"
                    class="absolute inset-0 object-cover"
                    src="https://api-ecom.duthanhduoc.com/images/bbea6d3e-e5b1-494f-ab16-02eece816d50.jpg"
                    style="opacity: 1;"
                  />
                </div>
                <div
                  class="slider relative mt-3 mb-1.5 grid grid-flow-col place-items-center gap-1.5 hidden"
                >
                  <div
                    class="w-full"
                  >
                    <div
                      class="relative w-full pb-[100%]"
                    >
                      <img
                        alt="image"
                        class="absolute inset-0"
                        src="https://api-ecom.duthanhduoc.com/images/bbea6d3e-e5b1-494f-ab16-02eece816d50.jpg"
                      />
                      <div
                        class="border-red-52c52 absolute inset-0 w-full border-2"
                      />
                    </div>
                  </div>
                  <button
                    class="absolute top-1/2 left-0 -translate-y-1/2 bg-black/20 text-white"
                  >
                    <svg
                      class="h-8 w-5"
                      fill="none"
                      stroke="currentColor"
                      stroke-width="3"
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M15.75 19.5 8.25 12l7.5-7.5"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                      />
                    </svg>
                  </button>
                  <button
                    class="absolute top-1/2 right-0 -translate-y-1/2 bg-black/20 text-white"
                  >
                    <svg
                      class="h-8 w-5"
                      fill="none"
                      stroke="currentColor"
                      stroke-width="3"
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="m8.25 4.5 7.5 7.5-7.5 7.5"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                      />
                    </svg>
                  </button>
                </div>
              </section>
              <section
                class="lg:col-span-7 lg:p-5"
              >
                <h1
                  class="text-xl"
                >
                  Điện Thoại Vsmart Active 3 6GB/64GB - Hàng Chính Hãng
                </h1>
                <div
                  class="group-price bg-gray-f9 flex items-center gap-2.5 px-5 py-4"
                >
                  <span
                    class="text-red-52c52 text-3xl first-letter:align-top first-letter:text-[26px]"
                  >
                    ₫3.190.000
                  </span>
                  <svg
                    fill="none"
                    height="21"
                    viewBox="0 0 20 21"
                    width="20"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M14.583 17.167c-2.533 0-4.585-2.067-4.585-4.616 0-2.55 2.052-4.616 4.585-4.616S19.167 10 19.167 12.55s-2.052 4.616-4.585 4.616m1.774-6.365a.5.5 0 0 1 .729-.017.51.51 0 0 1 .017.699l-2.633 2.923-.028.04a.513.513 0 0 1-.76.06l-1.402-1.33a.51.51 0 0 1-.01-.732.504.504 0 0 1 .701-.013l1.043.97z"
                      fill="#EE4D2D"
                      fill-rule="evenodd"
                    />
                    <path
                      d="M9.745 16.143H2.362a1.534 1.534 0 0 1-1.528-1.539v-3.462h.764a1.15 1.15 0 0 0 1.146-1.154 1.15 1.15 0 0 0-1.146-1.154H.834V5.372c0-.85.684-1.538 1.528-1.538h12.226c.844 0 1.529.688 1.529 1.538v1.325a6 6 0 0 0-1.147-.186V5.372a.383.383 0 0 0-.382-.384H6.443v1.718H5.425V4.988H2.362a.383.383 0 0 0-.382.384v2.34a2.304 2.304 0 0 1 1.91 2.276 2.304 2.304 0 0 1-1.91 2.277v2.34c0 .212.171.384.382.384h3.063v-1.512h1.018v1.512H9.08q.271.617.666 1.154"
                      fill="#EE4D2D"
                    />
                    <path
                      d="M5.425 9.579V7.732h1.018V9.58zm0 2.872v-1.847h1.018v1.847z"
                      fill="#EE4D2D"
                    />
                  </svg>
                  <span
                    class="text-gray-555/70 text-base line-through first-letter:align-top first-letter:text-sm"
                  >
                    ₫3.990.000
                  </span>
                  <span
                    class="text-normal bg-red-52c52 p-1 font-bold text-white uppercase"
                  >
                    20%
                     Giảm
                  </span>
                </div>
                <button
                  class="my-4 flex cursor-pointer items-center"
                >
                  Bảng quy đổi kích cỡ
                  <svg
                    class="size-5 text-gray-400"
                    fill="none"
                    stroke="currentColor"
                    stroke-width="1.5"
                    viewBox="0 0 24 24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m8.25 4.5 7.5 7.5-7.5 7.5"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                    />
                  </svg>
                </button>
                <div
                  class="group-quantity mt-4 flex items-center gap-3.5"
                >
                  <h2
                    class="w-[100px]"
                  >
                    Số lượng
                  </h2>
                  <div
                    class="flex items-center"
                  >
                    <button
                      aria-label="minus"
                      class="border-gray-ccc grid size-8 cursor-pointer place-items-center border disabled:bg-gray-300"
                    >
                      <svg
                        class="size-2.5"
                        fill="none"
                        stroke="currentColor"
                        stroke-width="1.5"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M5 12h14"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                        />
                      </svg>
                    </button>
                    <div
                      class="z-10"
                    >
                      <input
                        aria-live="assertive"
                        class=" border-gray-ccc text-red-52c52 focus-visible:ring-2 focus-visible:ring-black focus-visible:ring-offset-2 grid h-8 w-12.5 outline-none cursor-text place-items-center border-y"
                        type="text"
                        value="1"
                      />
                    </div>
                    <button
                      aria-label="plus"
                      class="border-gray-ccc grid size-8 cursor-pointer place-items-center border disabled:bg-gray-300"
                    >
                      <svg
                        class="size-2.5"
                        fill="none"
                        stroke="currentColor"
                        stroke-width="1.5"
                        viewBox="0 0 24 24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M12 4.5v15m7.5-7.5h-15"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                        />
                      </svg>
                    </button>
                  </div>
                  <div
                    class="text-gray-555 uppercase"
                  >
                    138
                     
                    Sản phẩm có sẵn
                  </div>
                </div>
                <div
                  class="mt-7.5 inline-flex items-center gap-3"
                >
                  <button
                    class="button"
                  >
                    Thêm vào giỏ hàng
                  </button>
                  <button
                    class="border border-black bg-white text-black px-2 py-4 text-nowrap"
                  >
                    Mua ngay
                  </button>
                </div>
              </section>
            </section>
          </div>
        </div>
      </main>
    </div>
    <section
      aria-atomic="false"
      aria-label="Notifications Alt+T"
      aria-live="polite"
      aria-relevant="additions text"
      class="Toastify"
    />
  </div>
</body>
`;
